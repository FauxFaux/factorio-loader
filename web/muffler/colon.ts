import { JFluid, JItem } from '../objects';
import { data } from '../datae';

export type Colon = string;

export function splitColon(colon: Colon): [string, string] {
  const splut = colon.split(':', 2);
  if (splut.length === 2) return splut as [string, string];
  throw new Error(`invalid colon ${colon}`);
}

export function fromColon(key: string): ['item', JItem] | ['fluid', JFluid] {
  if (key.startsWith('item:')) {
    return ['item', data.items[key.slice('item:'.length)]];
  }

  if (key.startsWith('fluid:')) {
    return ['fluid', data.fluids[key.slice('fluid:'.length)]];
  }

  throw new Error(`invalid colon'd name: "${key}"`);
}

export function tupleToColon([type, name]: [string, string]): Colon {
  return `${type}:${name}`;
}

export function objToColon(p: { type: string; name: string }): Colon {
  return `${p.type}:${p.name}`;
}

export function isRegular(colon: Colon): boolean {
  return colon.startsWith('item:') || colon.startsWith('fluid:');
}
