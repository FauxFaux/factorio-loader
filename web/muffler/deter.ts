export function sortByKeys<T>(obj: Record<string, T>): Record<string, T> {
  return Object.fromEntries(
    Object.entries(obj).sort(([a], [b]) => a.localeCompare(b)),
  );
}
