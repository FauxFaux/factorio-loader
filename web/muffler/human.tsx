export function humaniseNo(count: number): string {
  if (count > 1e6)
    return (
      (count / 1e6).toLocaleString('en', { maximumFractionDigits: 0 }) + 'M'
    );
  if (count > 1e3)
    return (
      (count / 1e3).toLocaleString('en', { maximumFractionDigits: 0 }) + 'k'
    );
  return count.toLocaleString('en', { maximumFractionDigits: 0 });
}

export function humanise(count: number | undefined) {
  if (count === undefined) return <abbr title="data absent">?</abbr>;
  if (count > 1e6)
    return (
      <abbr
        title={`${count.toLocaleString('en', { maximumFractionDigits: 0 })}`}
      >
        {(count / 1e6).toFixed() + 'M'}
      </abbr>
    );
  if (count > 1e3)
    return (
      <abbr
        title={`${count.toLocaleString('en', { maximumFractionDigits: 0 })}`}
      >
        {(count / 1e3).toFixed() + 'k'}
      </abbr>
    );
  if (count.toString() === count.toFixed()) {
    return <span>{count}</span>;
  }

  if (Math.abs(count) > 10) {
    return <abbr title={`${count.toPrecision(5)}`}>{count.toFixed()}</abbr>;
  }

  return <abbr title={`${count.toPrecision(5)}`}>{count.toPrecision(1)}</abbr>;
}
