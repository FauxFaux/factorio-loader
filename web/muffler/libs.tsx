import { JSX } from 'preact';
import { useState } from 'preact/hooks';
import { serializeError } from 'serialize-error';

import type { BrotliWasmType } from 'brotli-wasm';
import type Leaflet from 'leaflet';
import type * as ProcessMgmt from '../stubs/process-mgmt';

type Lib = 'brotli-wasm' | 'leaflet' | 'process-mgmt';

async function load(lib: Lib) {
  switch (lib) {
    case 'brotli-wasm':
      return import(/* webpackPrefetch: true */ 'brotli-wasm');
    case 'leaflet':
      return import(/* webpackPrefetch: true */ 'leaflet');
    case 'process-mgmt':
      return import(/* webpackPrefetch: true */ '../stubs/process-mgmt');
    default:
      return assertUnreachable(lib);
  }
}

export type LibState<T> = [JSX.Element, null] | [null, T];

export function useLib(lib: 'leaflet'): LibState<typeof Leaflet>;
export function useLib(lib: 'brotli-wasm'): LibState<BrotliWasmType>;
export function useLib(lib: 'process-mgmt'): LibState<typeof ProcessMgmt>;
export function useLib(lib: Lib): LibState<unknown> {
  const [state, setState] = useState({ lib: null, error: null } as {
    lib: unknown | null;
    error: Error | null;
  });
  if (state.error) {
    return [<>Loading failed: {serializeError(state.error)}</>, null];
  }
  if (state.lib) {
    return [null, state.lib];
  }
  load(lib)
    .then((lib) => {
      setState({ lib, error: null });
    })
    .catch((error) => {
      console.error(`Failed to load ${lib}`, error);
      setState({ lib: null, error });
    });

  return [<>Loading {lib}...</>, null];
}

function assertUnreachable(_value: never): never {
  throw new Error('Statement should be unreachable');
}
