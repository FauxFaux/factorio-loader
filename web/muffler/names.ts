import { data } from '../datae';
import { singularize } from 'inflection';

export function cleanupName(name: string) {
  const stripIcons = /\[[a-z0-9-]+=[a-z0-9-]+]/g;
  const normaliseSpace = /\s+/g;
  return name.replace(stripIcons, ' ').replace(normaliseSpace, ' ').trim();
}

export function compareWithoutIcons(a: string, b: string) {
  return cleanupName(a).localeCompare(cleanupName(b), 'en', {
    sensitivity: 'base',
  });
}

function strIEq(a: string, b: string): boolean {
  return (
    0 ===
    a.localeCompare(b, 'en', {
      sensitivity: 'base',
      usage: 'search',
      ignorePunctuation: true,
    })
  );
}

const labelLookupCache: Record<string, string> = {};

export function closelyMatchesItemName(label: string): string | null {
  if (Object.keys(labelLookupCache).length === 0) {
    for (const [name, item] of Object.entries(data.items)) {
      labelLookupCache[item.localised_name.toLowerCase()] = name;
    }
  }

  const quick = labelLookupCache[label.toLowerCase()];
  if (quick) return quick;

  switch (label.toLowerCase()) {
    // typo: inductor
    case 'air-core conductor':
      return 'inductor1';
  }
  for (const [name, item] of Object.entries(data.items)) {
    if (
      strIEq(label, item.localised_name) ||
      strIEq(singularize(label), item.localised_name)
    ) {
      return name;
    }
  }

  return null;
}

export function closelyMatches(label: string): string | null {
  let cleaned = cleanupName(label);
  let v = closelyMatchesItemName(cleaned);
  if (v) return v;
  v = closelyMatchesItemName(cleaned + ' barrel');
  if (v) return v;
  if (cleaned.startsWith('Auog ')) {
    v = closelyMatchesItemName(cleaned.slice('Auog '.length));
    if (v) return v;
  }

  return null;
}
