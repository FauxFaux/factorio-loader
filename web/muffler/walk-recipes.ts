import { computed, data } from '../datae';
import { splitColon, tupleToColon } from './colon';
import { JIngredient } from '../objects';

export const hiddenRequirements: Record<string, JIngredient> = {
  'caged-scrondrix-1': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-1a': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-2': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-2a': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-3': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-3a': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-4': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-4a': { colon: 'item:scrondrix', amount: 0 },
  'caged-scrondrix-5': { colon: 'item:scrondrix', amount: 0 },
  'scrondrix-cub-1': { colon: 'item:scrondrix', amount: 0 },
  'scrondrix-cub-2': { colon: 'item:scrondrix', amount: 0 },
  'scrondrix-cub-3': { colon: 'item:scrondrix', amount: 0 },
  'scrondrix-cub-4': { colon: 'item:scrondrix', amount: 0 },
  'scrondrix-mature-01': { colon: 'item:scrondrix', amount: 0 },
  'caged-dingrits1': { colon: 'item:dingrits', amount: 0 },
  'caged-dingrits2': { colon: 'item:dingrits', amount: 0 },
  'caged-dingrits3': { colon: 'item:dingrits', amount: 0 },
  'caged-dingrits4': { colon: 'item:dingrits', amount: 0 },
  'caged-dingrits5': { colon: 'item:dingrits', amount: 0 },
  'caged-dingrits6': { colon: 'item:dingrits', amount: 0 },
  'caged-dingrits7': { colon: 'item:dingrits', amount: 0 },
  'dingrits-mature-01': { colon: 'item:dingrits', amount: 0 },
};

export function recipeBan(name: string): boolean {
  return (
    name.endsWith('-barrel') ||
    name.endsWith('-pyvoid') ||
    name.endsWith('-pyvoid-fluid') ||
    name.endsWith('-pyvoid-gas')
  );
}

export function buildMaking() {
  const recipesMaking: Record<string, string[]> = {};
  for (const [name, recipe] of Object.entries(data.recipes.regular)) {
    if (recipeBan(name)) continue;
    for (const prod of recipe.products) {
      const colon = prod.colon;
      if (!recipesMaking[colon]) recipesMaking[colon] = [];
      recipesMaking[colon].push(name);
    }
  }
  return recipesMaking;
}

export function buildMissingIngredients(
  canMake: Set<string>,
  recipesMaking: Record<string, string[]>,
) {
  const missingIngredients: Record<string, number> = {};
  const countMissing = (name: string, context: string[] = []): number => {
    if (missingIngredients[name] === null) {
      // console.log('circular dependency', name, context);
      return 100;
    }
    if (missingIngredients[name] !== undefined) {
      return missingIngredients[name];
    }
    missingIngredients[name] = null as any;
    const recipe = data.recipes.regular[name];
    if (!recipe) {
      return 10;
    }
    let missing = 0;
    const products = new Set(...recipe.products.map((p) => p.colon));
    for (const ing of (recipe.ingredients ?? []).concat(
      hiddenRequirements[name] ?? [],
    )) {
      const colon = ing.colon;
      if (products.has(colon)) continue;
      if (canMake.has(colon)) continue;
      if (!recipesMaking[colon]) {
        const [type, sub] = splitColon(colon);
        if (type === 'item') {
          const fluid = computed.barrelFluid[sub];
          if (fluid) {
            missing += Math.min(
              ...recipesMaking[tupleToColon(['fluid', fluid])].map((name) =>
                countMissing(name, context.concat(colon + ' (fluid)')),
              ),
            );
          } else {
            missing += 20;
          }
        } else {
          missing += 20;
        }
        continue;
      }
      missing +=
        Math.min(
          ...recipesMaking[colon].map((name) =>
            countMissing(name, context.concat(colon)),
          ),
        ) + 1;
    }
    missingIngredients[name] = missing;
    return missing;
  };

  // for (const name of Object.keys(data.recipes.regular)) {
  //   if (recipeBan(name)) continue;
  //   countMissing(name);
  countMissing('arthurian-egg-01');
  // }
  return missingIngredients;
}
