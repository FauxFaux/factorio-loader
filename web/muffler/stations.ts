import { Stop } from '../../scripts/load-recs';
import { Colon, fromColon, tupleToColon } from './colon';
import { data } from '../datae';
import { cleanupName, closelyMatches } from './names';

export function isProvideStation(name: string): boolean {
  return !!/\bProvide\b/i.exec(name);
}

export function provideStationPurpose(name: string): Set<string> {
  const matches = new Set<string>();
  for (const ma of name.matchAll(/\[item=([a-z0-9-]+)]/g)) {
    const item = ma[1];
    if (item in data.items) {
      matches.add(item);
    }
  }

  if (matches.size) return matches;
  name = cleanupName(name);
  let ma = name.match(/^([a-zA-Z0-9 /-]+) Provide/i);
  if (ma) {
    for (const part of ma[1].split('/')) {
      const guess = closelyMatches(part);
      if (guess) {
        matches.add(guess);
      }
    }
  }
  return matches;
}

export type Stat = readonly [string, Stop];

export function stations(): Stat[] {
  return Object.entries(data.doc).flatMap(([loc, brick]) =>
    brick.stop.map((stop) => [loc, stop] as const),
  );
}

export function itemMap(stop: Stop): Record<string, number> {
  return Object.fromEntries(
    stop.items
      .filter(([kind]) => kind === 'item')
      .map(([, name, value]) => [name, value] as const),
  );
}

export function colonMapItems(stop: Stop): Record<Colon, number> {
  return Object.fromEntries(
    stop.items
      .filter(([kind]) => kind !== 'virtual')
      .map(
        ([type, name, value]) => [tupleToColon([type, name]), value] as const,
      ),
  );
}
export function colonMapCombinator(stop: Stop): Record<Colon, number> {
  return Object.fromEntries(
    stop.combinator.map(
      ([type, name, value]) => [tupleToColon([type, name]), value] as const,
    ),
  );
}

interface LtnSettings {
  'ltn-provider-stack-threshold'?: number;
  'ltn-provider-threshold'?: number;

  // there are other options here, they're present in the data but I have not mapped them
}

export function settingsMap(stop: Stop): LtnSettings {
  return Object.fromEntries(
    stop.settings
      .filter(([kind, name]) => kind === 'virtual' && name.startsWith('ltn-'))
      .map(([, name, value]) => [name, value] as const),
  );
}

export function ltnMinTransfer(colon: Colon, settings: LtnSettings) {
  const [type, item] = fromColon(colon);
  const expectedByStack =
    type === 'item'
      ? item.stack_size * (settings['ltn-provider-stack-threshold'] ?? 10)
      : 0;
  const expectedByCount = settings['ltn-provider-threshold'] ?? 1;
  return Math.max(expectedByStack, expectedByCount);
}
