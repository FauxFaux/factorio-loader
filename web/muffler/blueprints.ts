import { atob, btoa } from 'abab';
import pako from 'pako';
import _chunk from 'lodash/chunk';
import { Colon, splitColon, tupleToColon } from './colon';
import { makeUpRecipe, RecipeName } from './walk-recipes';
import { Factory, FactoryClass } from '../datae';

export interface Blueprint {
  entities?: Entity[];
  tiles?: { name: string; position?: unknown }[];

  icons?: unknown[];

  // from the map version
  version: number;

  item: 'blueprint';

  // incomplete
}

export interface Entity {
  entity_number?: number;

  // presumably always an item; this is *not* colon'd
  name: string;
  inventory?: unknown;
  // e.g. { speed_module: 6 } for factories, not an inventory apparently
  items?: Record<string, number>;
  position?: unknown;
  recipe?: string;
  direction?: number;
  // massively over-specified
  control_behavior?: {
    logistic_condition: {
      first_signal: {
        type: 'item';
        name: string;
      };
      constant: number;
      comparator: '<';
    };
    connect_to_logistic_network: true;
  };
  // e.g. power poles
  neighbours?: number[];

  request_filters?: { name: string; count: number; index?: number }[];
}

export function decode(data: string): Blueprint {
  if (data[0] !== '0') throw new Error(`unsupported version ${data[0]}`);
  const un64 = Uint8Array.from(atob(data.slice(1))!, (c) => c.charCodeAt(0));
  const obj = JSON.parse(pako.inflate(un64, { to: 'string' }));
  const keys = Object.keys(obj);
  if (keys.length !== 1 || keys[0] !== 'blueprint')
    throw new Error(`invalid top level: ${keys}`);
  return obj.blueprint as Blueprint;
}

export function encode(blueprint: Blueprint): string {
  const obj = {
    blueprint,
  };
  const un64 = pako.deflate(JSON.stringify(obj), { level: 9 });
  return '0' + btoa(String.fromCharCode(...un64));
}

export function enumerate(input: Blueprint): Record<Colon, number> {
  const result: Record<Colon, number> = {};
  for (const entity of input.entities ?? []) {
    const colon = tupleToColon(['item', entity.name]);
    result[colon] = (result[colon] ?? 0) + 1;
  }
  return result;
}

export function buildRequestFilters(input: Record<Colon, number>) {
  return Object.entries(input).map(([colon, count], i) => {
    const [, name] = splitColon(colon);
    return {
      count,
      index: i + 1,
      name,
    };
  });
}

export function toChest(
  reference: Blueprint,
  input: Record<Colon, number>,
  chestName = 'logistic-chest-requester',
): Blueprint {
  const justChest = {
    index: 1,
    signal: {
      name: chestName,
      type: 'item',
    },
  };
  return {
    entities: [
      {
        entity_number: 1,
        name: chestName,
        position: {
          x: 0,
          y: 0,
        },
        request_filters: buildRequestFilters(input),
      },
    ],
    icons: [justChest],
    item: 'blueprint',
    version: input.version,
  };
}

export function toBlueprint(entities: Entity[]): Blueprint {
  return {
    entities: entities.map((entity, entity_number) => ({
      ...entity,
      entity_number: entity_number + 1,
    })),
    item: 'blueprint',
    version: 281479275151360,
  };
}

function ingredientMap(recp: string): Record<Colon, number> {
  if (!recp) return {};
  return Object.fromEntries(
    makeUpRecipe(recp)?.ingredients?.map(
      (ing) => [ing.colon, ing.amount] as const,
    ) ?? [],
  );
}

function mergeReq(
  leftReq: Record<string, number>,
  rightReq: Record<string, number>,
): Record<string, number> {
  const result: Record<string, number> = {};
  for (const [name, amount] of Object.entries(leftReq)) {
    result[name] = (result[name] ?? 0) + amount;
  }
  for (const [name, amount] of Object.entries(rightReq)) {
    result[name] = (result[name] ?? 0) + amount;
  }
  return result;
}

function roundUpFour(n: number) {
  return Math.ceil(n / 4) * 4;
}

function roundUpReq(req: Record<string, number>): Record<string, number> {
  return Object.fromEntries(
    Object.entries(req).map(
      ([name, amount]) => [name, roundUpFour((amount + 1) * 1.1)] as const,
    ),
  );
}

function onlyProduct(recp: string) {
  const recpD = makeUpRecipe(recp);
  const products = recpD?.products ?? [];
  if (1 !== products.length)
    throw new Error(
      `need exactly one product for ${recp}: ${JSON.stringify(recpD)}`,
    );
  return splitColon(products[0].colon)[1];
}

export function mallAssemblers(recipes: RecipeName[][]): Entity[] {
  let x = 0;
  let y = 0;
  const entities: Entity[] = [];
  let lastPole: number | undefined;
  for (const [leftStack, rightStack] of _chunk(recipes, 2)) {
    lastPole = undefined;
    for (let stackItem = 0; stackItem < leftStack.length; ++stackItem) {
      const leftRecp: RecipeName | undefined = leftStack[stackItem];
      const rightRecp: RecipeName | undefined = rightStack?.[stackItem];
      const leftReq = ingredientMap(leftRecp);
      const rightReq = ingredientMap(rightRecp);
      const req = roundUpReq(mergeReq(leftReq, rightReq));

      if (leftRecp) {
        entities.push({
          name: 'assembling-machine-2',
          position: { x, y },
          recipe: leftRecp,
        });
        entities.push({
          name: 'stack-inserter',
          position: { x: x + 2, y },
          // inserting to the right
          direction: 2,
        });
        entities.push({
          name: 'inserter',
          position: { x: x + 2, y: y - 1 },
          // inserting to the left
          direction: 6,
          control_behavior: {
            logistic_condition: {
              first_signal: {
                type: 'item',
                name: onlyProduct(leftRecp),
              },
              constant: 5,
              comparator: '<',
            },
            connect_to_logistic_network: true,
          },
        });
      }
      if (Object.entries(req).length !== 0) {
        entities.push({
          name: 'logistic-chest-requester',
          request_filters: buildRequestFilters(req),
          position: { x: x + 3, y },
        });
      }
      if (rightRecp) {
        entities.push({
          name: 'stack-inserter',
          position: { x: x + 4, y },
          // inserting to the left
          direction: 6,
        });
        entities.push({
          name: 'assembling-machine-2',
          position: { x: x + 6, y },
          recipe: rightRecp,
        });
        entities.push({
          name: 'inserter',
          position: { x: x + 4, y: y - 1 },
          // inserting to the right
          direction: 2,
          control_behavior: {
            logistic_condition: {
              first_signal: {
                type: 'item',
                name: onlyProduct(rightRecp),
              },
              constant: 5,
              comparator: '<',
            },
            connect_to_logistic_network: true,
          },
        });
      }
      if (leftRecp || rightRecp) {
        entities.push({
          name: 'logistic-chest-passive-provider',
          position: { x: x + 3, y: y - 1 },
        });
      }
      entities.push({
        name: 'small-electric-pole',
        position: { x: x + 3, y: y + 1 },
        neighbours: lastPole ? [lastPole] : undefined,
      });
      // XXX sensitive to entities order
      lastPole = entities.length;
      y += 3;
    }
    y = 0;
    x += 9;
  }

  return entities;
}

export function stripProducer(p: Factory): FactoryClass {
  return p
    .replace(/-?mk0\d|-\d$/, '')
    .replace('assembling-machine', 'automated-factory')
    .replace(
      /advanced-foundry|electric-furnace|steel-furnace|stone-furnace/,
      'furnace',
    )
    .replace('pumpjack-hightech', 'pumpjack');
}

export function stripProducers(producers: Factory[]): FactoryClass[] {
  return [...new Set(producers.map((p) => stripProducer(p)))].sort();
}
