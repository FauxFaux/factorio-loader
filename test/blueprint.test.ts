import { decode } from '../web/muffler/blueprints';

test('blueprint decode', () => {
  const train =
    '0eNqdUk1rwzAM/StBZyfkY26awE7bpYfddiuluIlITR07OE66UvLfJzctK3QbbPhgS356T5bfGXZqwM5K7aA8g6yM7qFcn6GXjRbK59ypQyhBOmyBgRatj/YuVKYyrXFyRJgYSF3jB5TJtGGA2kkncSa6BKetHtodWgL8QMGgMz1VGe01iSlMM87gRIesiKOEL1JOMsZKIhQzLp7YA396z9+gRiur8Cgawv8msUyjYhEXWf4oEnH/upEyxhJWD0p9o5v9Tzfnf9Kl4fbVHutBXaf7NUEfJ3f38y/e2GA9SusGocL5Y5+V02GNnXGb4NVv1OVRSLclA9SXXmd+Yu+Exe3VBMYS7mYILSoSpiH4nKwOVJHF8bTxi81+Ke/sxUCJHZKl4EXYxgTvVkgdvB2C1YruRrT9pdN0mTzlRZrzhCfZIp6mTwVq34Y=';
  const smallPartsRequest =
    '0eNqFkEsOwjAMRO/idYqSQvjkKqhCpRiwlCalSRFVlbvjlI/YsRrZGr0Ze4KjHbDryUUwE1DjXQCznyDQxdU27+LYIRigiC0IcHWbJ+svFCI1RXPFEIsebwMr9pAEkDvhA4xKlQB0kSLhizkP48EN7ZGdRv2nCeh8YIB3uQlDy4UWMLLKheaot/NwJsv2V8on/ksPbW1t0dV9DIVUzGz8kM/VUqYql5xPMz+fEHBn2pxabtVqsys3Wmm1XMuUnj6OZkc=';
  expect(decode(train)).toMatchSnapshot();
  expect(decode(smallPartsRequest)).toMatchSnapshot();
});
