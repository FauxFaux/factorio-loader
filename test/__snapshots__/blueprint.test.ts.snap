// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blueprint decode 1`] = `
{
  "blueprint": {
    "entities": [
      {
        "entity_number": 1,
        "name": "ht-locomotive",
        "orientation": 0,
        "position": {
          "x": -235,
          "y": -390.15625,
        },
      },
      {
        "entity_number": 2,
        "inventory": null,
        "name": "ht-generic-wagon",
        "orientation": 0.5,
        "position": {
          "x": -235,
          "y": -382.9609375,
        },
      },
      {
        "entity_number": 3,
        "inventory": null,
        "name": "ht-generic-wagon",
        "orientation": 0.5,
        "position": {
          "x": -235,
          "y": -375.9609375,
        },
      },
    ],
    "icons": [
      {
        "index": 1,
        "signal": {
          "name": "ht-locomotive",
          "type": "item",
        },
      },
    ],
    "item": "blueprint",
    "label": "Cargo Train Mk II",
    "schedules": [
      {
        "locomotives": [
          1,
        ],
        "schedule": [
          {
            "station": "[virtual-signal=ltn-depot] Depot",
            "wait_conditions": [
              {
                "compare_type": "or",
                "ticks": 300,
                "type": "inactivity",
              },
            ],
          },
        ],
      },
    ],
    "version": 281479275151360,
  },
}
`;

exports[`blueprint decode 2`] = `
{
  "blueprint": {
    "entities": [
      {
        "entity_number": 1,
        "name": "logistic-chest-requester",
        "position": {
          "x": 2.5,
          "y": 20.5,
        },
        "request_filters": [
          {
            "count": 500,
            "index": 1,
            "name": "small-parts-01",
          },
        ],
      },
    ],
    "icons": [
      {
        "index": 1,
        "signal": {
          "name": "logistic-chest-requester",
          "type": "item",
        },
      },
    ],
    "item": "blueprint",
    "version": 281479275151360,
  },
}
`;
